Use 'updated_at' to prevent overwriting new data with old data.

psvet opened the pull request:
#5599
- refactored `models/base/index#edit` to return a transaction
- added `safeToSave` validation method to check `updated_at` in client and DB objects
- if client-side `updated_at` is behind DB, and other changes have been made, throw `409 Conflict` error
- added `deep-diff` module for calculating changes between objects

Marked WIP because differences in tags currently aren't tracked when updating posts. Input regarding next steps will be appreciated!

ErisDS commented:
Apologies for the slow response here - this is looking pretty darned cool!

I've taken it for a quick spin and the handling of the error is going to need a little love at some point as at the moment if I trigger the error, what I see is this, which isn't ideal:

![](http://puu.sh/kHUIc.png)

I think you've got a relatively sensible approach to which values to check included in the PR - it may need tweaking as we test it out and figure out what feels 'right'.

How can I help this move forward, e.g. with tracking the tags?

psvet commented:
@ErisDS, can you tell me how you produced that error? I haven't seen that vague message since adding this functionality, which makes me think I'm perhaps working on a different case than you tested.

Just to clarify, my understanding was that the error should be thrown if a user (or multiple users) happens to be working on the same post from different locations (computers, browsers, etc.) and attempts to save without having refreshed the page since a previous save. I have this working on my end with the correct error message. Is there another case I should consider?

And regarding the tags: I haven't been able to look at it for a few days, but as far as I can tell it's just a matter of updating the Bookshelf query to include the `withRelated` option. The hang-up was figuring out exactly how and where to add it and keep track of it. Since it's post-specific it'll need to be included higher up the line, before the transaction in `models/base`.

If you have any suggestions off the top of your head, I'll be happy to hear them. I'll get back to work on this as soon as I get my old computer up and running again, because I left my main one behind in *Russia*, of all places. Won't have it until Monday or so...

ErisDS commented:
I was testing using two tabs in chrome, when I was getting that error, but trying again now I get the proper error message.

The relationship & management of saving tags when storing a post is one of those gnarly areas, so perhaps it's not worth thinking about for now.

If we could get this branch updated that would be great. The error message is also going to need a little bit of consideration before this gets in front of users, although that can be looked at later:

![](http://puu.sh/kSooy.png)

cc @JohnONolan - this change will prevent users from accidentally overwriting their posts by having multiple tags open, and saving the tab with an older version. The error message is currently in an alert because something went wrong, and you can't continue until you fix it. 

A friendlier version of the text might be something like this:

 > Uh-oh. We already have a newer version of this post saved. To prevent losing your text, please copy your changes somewhere else and then refresh this page.

In the majority of cases, this should simply catch people doing the multiple-tab thing by accident - I know I've done it. For blogs with multiple users, it would also be possible to get this if someone else has the post open and is editing it.

psvet commented:
Yeah, I was thinking similarly. I know I can hack something together to make the tags *work*, but it really looks like a fairly substantial refactor would be necessary to make an ideal solution possible. I'll be happy to work on it more if/when you think it's more worthwhile. 

Update incoming, with the friendlier message. :+1:

psvet committed:
Use 'updated_at' to prevent overwriting new data with old data

Closes #5599
- refactored `models/base/index#edit` to return a transaction
- added `safeToSave` validation method to check `updated_at` in client and DB objects
- if client-side `updated_at` is behind DB, and other changes have been made, throw `409 Conflict` error
- added `deep-diff` module for calculating changes between objects

kirrg001 commented:
I like this feature, but this PR is pretty old!
As long as the branch does not get deleted, we can re-open the PR anytime 👍 

Thanks for your work @psvet
Closing.
